### PR TITLE
Changes for adding automatic classes for any analysis form

### PR DESCRIPTION
**Pull request by xavijam**
Related ticket: #12095.

Basically, we need to add track classes in all analysis form editor components. With that, we could check if they are working as expected at UI perspective.

## Pending things
- [x] Decide the good way to add them in the current analysis form "workflow".
- [x] Check all backbone form editors will have this class added when needed (check `_setOptions` method is called in their respective initialize'rs').
- [x] Check the behavior for the nested forms.
- [x] Add necessary tests.

## Steps to test it
We have changed the way we set the options in any editor component, so I encourage you to test all components in different contexts. One, withing an analysis form (with track class), and another out of the analysis world. I've created a table with all Backbone.Form.editors available in Builder, providing a place to check in an analysis and another out of it.

| Reviewed | Name | Analysis to find it (with track class)  | Place to find it (without track class)  |
|---|---|---|---|
| [x] | CodeEditor  |  Geocoding by street address (advanced mode)  | _x_  | 
| [-] | DataObservatoryDropdown  | Data-observatory :D  | _x_  | 
| [x] | DateTime  | _x_  | Editing a feature  | 
| [x] | Enabler  | Connect with lines  | _x_  | 
| [x] | EnablerEditor  | Centroid  | Editing a widget, in the form  | 
| [x] | Fill  | _x_  | Style a layer  | 
| [-] | List  | _x_  | _x_  |
| [x] | MultiSelect  | Markov  | _x_  |
| [x] | NodeDataset  | _Any analysis_  | _x_  |
| [x] | Number  | Area of influence  | Export image form  |
| :boom: | Operators  | Intersection  | Aggregation options when style is aggregated |
| [x] | Radio  | Merge  | Widget form view |
| [x] | Select  | Closest  | Style form view, selecting a label  |
| [-] | Slider  | _x_  | _x_  |
| [x] | SortableList  | _x_  | Customizing a legend  |
| [x] | Suggest  | _x_  | Editing a feature |
| [x] | Taglist  | _x_  | Map metadata form view (within the modal) |
| [x] | Text  | Filter analysis | Widget edit form view |
| [x] | TextArea  | _x_ | Formula widget edit form view |
| [x] | Toggle  | Georeference by street address (advance mode is the toggle) | _x_ |

**Notes**
There are several Backbone.Forms editors not migrated yet: `[Checkbox, Checkboxes, Date, Hidden, Password]`. And also there are some those don't need anything related with classes `[NestedModel, Base, Object,...]`.

**Legend**
x -> There is no place using that editor.

- [x] Apart from that, we have to check that the apply button in any analysis form view has the proper `track` class.

### CODE REVIEW

File: lib/assets/core/javascripts/cartodb3/components/form-components/field.js [hunk 1]
```diff
@@ -16,6 +22,37 @@ Backbone.Form.Field = Backbone.Form.Field.extend({
...
+  * OVERWRITTEN the creation of the full field schema, merging defaults etc.
+  *
+  * @param {Object|String} schema
+  *
+  * @return {Object}
+  */
+  createSchema: function (schema) {
+    var editorType = schema.type;
+    schema = this.constructor.__super__.createSchema.apply(this, arguments);
+    schema.editorType = editorType;
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/field.js]: why do we need to set `editorType` and `type` pointing to the same value? is not `schema.type` enough?

**Comment by nobuti** [comment]
I like the approach, I think is the way to go. The only concern I have it's the case of nested forms, are we going to add some kind of index? For instance, in DO, the user can add multiple forms in the manner of measurements. Will all the common fields have the same css class?

File: lib/assets/core/javascripts/cartodb3/components/form-components/field.js (hunk 1 above)
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/field.js]: It replaces schema.type by a Function.

**Comment by xavijam** [comment]
I haven't thought about Nested ones. Let me add it to the list in order to check them.

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/editor-helpers-extend.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+var _ = require('underscore');
+
+module.exports = {
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/editor-helpers-extend.js]: Te resistes a crear un mixin... 😂

Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/editor-helpers-extend.js]: Yep :joy:

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/text.js [hunk 3]
```diff
@@ -1,12 +1,24 @@
...
   initialize: function (opts) {
-    Backbone.Form.editors.Base.prototype._setOptions.call(this, opts); // Options
-    this.constructor.__super__.initialize.apply(this, arguments);
+    Backbone.Form.editors.Base.prototype.initialize.call(this, opts);
+    EditorHelpers.setOptions(this, opts);
+
+    var schema = this.schema;
+
+    // Allow customising text type (email, phone etc.) for HTML5 browsers
+    var type = 'text';
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/text.js]: nice!

File: lib/assets/core/test/spec/cartodb3/components/form-components/editors/slider.spec.js [hunk 4]
```diff
@@ -84,7 +87,7 @@ describe('components/form-components/editors/slider', function () {
     describe('exceptions', function () {
       it('should throw exception when the values and labels differ', function () {
         var wrapper = function () {
-          new Backbone.Form.editors.Slider({ // eslint-disable-line no-new
+          createViewFn({ // eslint-disable-line no-new
```
Comment (nobuti) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/slider.spec.js]: lint comment can be deleted

**Comment by nobuti** [comment]
I've found things, not sure if they are issues:
- The enabler editor doesn't transmit the tracking class to the inner editor. For instance, we are adding tracking classes to all select dropdowns, not sure if the inner ones should have the same class or not. 
screen shot 2017-05-15 at 07 59 01
screen shot 2017-05-15 at 07 59 36

However, this does happen in the other enabler editor:

screen shot 2017-05-15 at 07 49 04

screen shot 2017-05-15 at 07 56 09

- I can't find the track class in the CodeEditor:

screen shot 2017-05-15 at 08 16 25

Other than that, it's pretty cool!

**Comment by nobuti** [comment]
Moving back to development.

**Comment by nobuti** [comment]
Fixed, @xavijam could you take a look to my last (util) two commits?

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/code-editor.js [hunk 5]
```diff
@@ -56,6 +56,16 @@ Backbone.Form.editors.CodeEditor = Backbone.Form.editors.TextArea.extend({
     this.$el.replaceWith($codeMirrorEl);
     this.setElement($codeMirrorEl);
     this.$el.addClass('CodeMirror-formInput');
+    // The default el is replace it with another dom node
+    // we should add tracking class manually again
+    this._addTrackingClass();
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/code-editor.js]: Does it pass the generic tracking class spec for all editos?

Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/code-editor.js]: Basically [here](https://github.com/CartoDB/cartodb/pull/12106/files#diff-76a877fe9d26f2bf5b4a2324fdfd4e4c)?

**Comment by xavijam** [comment]
Fine by me BTW 💋

**Comment by xavijam** [comment]
Deploying it in ded03.

**Comment by xavijam** [comment]
The data-observatory-dropdown can't be used right now until new UI is released, isn't it @nobuti?

**Comment by nobuti** [comment]
Yes, that's correct. I'll check it out once it's merged, we have a custom
form in that branch too that probably we need to fine tune as well.

**Comment by xavijam** [comment]
Found a problem, operators editor is not working. Finishing acceptance before moving to development.

**Comment by xavijam** [comment]
Turns out that error is already in production (https://github.com/CartoDB/cartodb/issues/12133). So, not stopping this PR. Merging and checking classes in production @noguerol and @saleiva .

**Pull request merged**